criu/cr-check.c

```diff
@@ -1551,3 +1553,125 @@ static char *feature_name(int (*func)(void))
 	}
 	return NULL;
 }
+
+#ifndef CAP_CHECKPOINT_RESTORE
+#define CAP_CHECKPOINT_RESTORE 40
+#endif
+
+static bool has_capability(int cap, struct vfs_cap_data *cap_data)
+{
+	int mask = CAP_TO_MASK(cap);
+	int index = CAP_TO_INDEX(cap);
+	u32 inheritable;
+	u32 permitted;
+
+	inheritable = le32toh(cap_data->data[index].inheritable);
+	permitted = le32toh(cap_data->data[index].permitted);
+
+	if (!(mask & permitted)) {
+		pr_debug("Capability %d not permitted\n", cap);
+		return false;
+	}
+
+	if (!(mask & inheritable)) {
+		pr_debug("Capability %d not inheritable\n", cap);
+		return false;
+	}
+
+	return true;
+}
+
+static int pr_set_dumpable(int value)
+{
+	int ret = prctl(PR_SET_DUMPABLE, value, 0, 0, 0);
+	if (ret < 0)
+		pr_perror("Unable to set PR_SET_DUMPABLE");
+	return ret;
+}
+
+int check_uid(void)
+{
+	const char *attribute = "security.capability";
+	const char *criu = "/proc/self/exe";
+	uid_t uid = geteuid();
+	int exit_code = -1;
+	char *caps = NULL;
+	u32 magic;
+	int ret;
+	int len;
+	int fd;
+
+	if (uid == 0)
+		return 0;
+
+	/*
+	 * If running as non-root CRIU either needs CAP_SYS_ADMIN or
+	 * CAP_CHECKPOINT_RESTORE and CAP_PTRACE.
+	 */
+
+	fd = open(criu, O_RDONLY);
```

Comment:
criu already should have a helper to test whether or not it is running a user namespace. If not, what should work is sm like:
```c
inline static bool in_userns(void)
{
	__do_fclose FILE *f = NULL;
	uid_t user, host, count;
	int ret;

	/* Now: are we in a user namespace? Because then we're also
	 * unprivileged.
	 */
	f = fopen("/proc/self/uid_map", "re");
	if (!f)
		return false;

	ret = fscanf(f, "%u %u %u", &user, &host, &count);
	if (ret != 3)
		return false;

	return user != 0 || host != 0 || count != UINT32_MAX;
}
```